packages/block-library/src/link-preview/edit.js

```diff
@@ -0,0 +1,128 @@
+/**
+ * WordPress dependencies
+ */
+import { __, _x } from '@wordpress/i18n';
+import { useState, useEffect } from '@wordpress/element';
+import { useBlockProps, BlockControls } from '@wordpress/block-editor';
+import { __experimentalFetchUrlData } from '@wordpress/core-data';
+import {
+	Placeholder,
+	Spinner,
+	Button,
+	ToolbarGroup,
+	ToolbarButton,
+} from '@wordpress/components';
+import { link, edit } from '@wordpress/icons';
+
+export default function LinkPreviewEdit( props ) {
+	const {
+		attributes: { url, title, icon, image },
+		setAttributes,
+	} = props;
+	const [ isFetching, setIsFetching ] = useState( false );
+	const [ isEditingUrl, setIsEditingUrl ] = useState( ! url );
+	const [ urlValue, setURLValue ] = useState( '' );
+
+	const blockProps = useBlockProps( {
+		href: url,
+		className: image ? 'has-image' : undefined,
+		onClick: isEditingUrl
+			? undefined
+			: ( event ) => {
+					event.preventDefault();
+			  },
+	} );
+
+	useEffect( () => {
+		if ( url && ! title && ! icon && ! image ) {
+			setIsFetching( true );
+			__experimentalFetchUrlData( url )
+				.then( ( data ) => {
+					setAttributes( data );
+				} )
+				.finally( () => {
+					setIsFetching( false );
+				} );
+		}
+	}, [] );
+
+	if ( isEditingUrl ) {
+		return (
+			<div { ...blockProps }>
+				<Placeholder
+					icon={ link }
+					label={ __( 'URL' ) }
+					instructions={ __(
+						'Paste a link to the content you want to display on your site.'
+					) }
+				>
+					<form
+						onSubmit={ () => {
+							setAttributes( {
+								url: urlValue,
+								title: '',
+								icon: '',
+								image: '',
+								description: '',
+							} );
+							setIsFetching( true );
+							__experimentalFetchUrlData( urlValue )
```

Comment:
We may not need this if we put the effectful code in the `useEffect` and have a single place to fetch. It also makes the block attributes the single source of truth for the "link". In this case we're using the temporary local state variable as the url.

Reply:
I removed the effect. We shouldn't be doing any fetching that doesn't come from a user interaction since we're saving the content as block attributes. If there's no url or title, we automatically switch to edit mode.

Reply:
Restored the effect so it can be transformed to from an embed block